crates/sui-graphql-rpc/src/server/data_provider.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// For testing, use existing RPC as data source
+
+use std::str::FromStr;
+
+use crate::types::address::Address;
+use crate::types::balance::Balance;
+use crate::types::base64::Base64;
+use crate::types::big_int::BigInt;
+use crate::types::transaction_block::TransactionBlock;
+use crate::types::{object::Object, sui_address::SuiAddress};
+use async_graphql::*;
+use sui_json_rpc_types::{
+    SuiObjectDataOptions, SuiPastObjectResponse, SuiTransactionBlockResponseOptions,
+};
+use sui_sdk::types::base_types::ObjectID;
+use sui_sdk::types::digests::TransactionDigest;
+use sui_sdk::SuiClient;
+
+pub(crate) async fn fetch_obj(
+    cl: &SuiClient,
+    address: SuiAddress,
+    version: Option<u64>,
+) -> Result<Option<Object>> {
+    let oid: ObjectID = address.to_array().as_slice().try_into()?;
+    let opts = SuiObjectDataOptions::full_content();
+
+    let g = match version {
+        Some(v) => match cl
+            .read_api()
+            .try_get_parsed_past_object(oid, v.into(), opts)
+            .await?
+        {
+            SuiPastObjectResponse::VersionFound(x) => x,
+            _ => return Ok(None),
+        },
+        None => {
+            let val = cl.read_api().get_object_with_options(oid, opts).await?;
+            if val.error.is_some() || val.data.is_none() {
+                return Ok(None);
+            }
+            val.data.unwrap()
+        }
+    };
+    Ok(Some(convert_obj(g)))
+}
+
+pub(crate) async fn fetch_balance(
+    cl: &SuiClient,
+    address: &SuiAddress,
+    type_: Option<String>,
+) -> Result<Balance> {
+    let b = cl
+        .coin_read_api()
+        .get_balance(address.to_array().as_slice().try_into()?, type_)
+        .await?;
+    Ok(convert_bal(b))
+}
+
+fn convert_obj(s: sui_json_rpc_types::SuiObjectData) -> Object {
+    Object {
+        version: s.version.into(),
+        digest: s.digest.to_string(),
+        storage_rebate: s.storage_rebate,
+        address: SuiAddress::from_array(**s.object_id),
+        owner: s
+            .owner
+            .unwrap()
+            .get_owner_address()
+            .map(|x| SuiAddress::from_array(x.to_inner()))
+            .ok(),
+        bcs: Some(Base64::from(&bcs::to_bytes(&s.bcs).unwrap())), // TODO: is this correct?
```

Comment:
I don't think this is right for a couple of reasons:

- `s.bcs` is an `Option<SuiRawData>` -- we don't want to serialize the option part or the data around the bcs bytes within `SuiRawData` -- these are all `sui-json-rpc` implementation details.
- The relevant data within `SuiRawData` is already BCS serialized, so we shouldn't need to call `to_bytes` on it.

The actual implementation is quite nasty, it looks something like this:

```rust
bcs: s.bcs.map(|raw| Base64::from(match raw {
    SuiRawData::Package(raw_package) => {
        &bcs::to_bytes(&raw_package).unwrap()
    },

    SuiRawData::MoveObject(raw_object) => {
        &raw_object.bcs_bytes
    },
}))
```